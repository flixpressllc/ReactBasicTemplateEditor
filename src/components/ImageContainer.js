import React from 'react';
import {SortableContainer, SortableElement, SortableHandle, arrayMove} from 'react-sortable-hoc';
import CaptionInput from './CaptionInput';
import Modal from './lib/Modal';
import ImageDropDown from './ImageDropDown';
import { THUMBNAIL_URL_PREFIX } from '../stores/app-settings';
import { registerDataType } from '../utils/globalContainerConcerns';
import { clone, toType, forceArray } from 'happy-helpers';
import TemplateSpecificationsStore from '../stores/TemplateSpecificationsStore';
import { disableTextSelectionOnTheWholeBody, enableTextSelectionOnTheWholeBody } from '../utils/browser-specific-hacks';
import * as ContainerActions from '../actions/ContainerActions';
import PT from 'prop-types';

import './ImageContainer.scss';

const DATA_TYPE_NAME = 'userImageChooser';

function prepCaptions(imageObject) {
  imageObject.captions = forceArray(imageObject.captions);

  imageObject.captions = imageObject.captions.map((caption) => {
    return {
      label: caption.label,
      value: caption.value
    };
  })
  return imageObject;
}

function prepDropDowns(imageObject) {
  imageObject.dropDowns = forceArray(imageObject.dropDowns);

  imageObject.dropDowns = imageObject.dropDowns.map((dropDown) => {
    return {
      label: dropDown.label,
      value: dropDown.value
    };
  })
  return imageObject;
}

// This won't really be a string. That's okay, though.
export function toRenderString (imageChooserObj) {
  let renderValue = clone(imageChooserObj.containedImages).map(imgObj => {
    let newImageObj = prepCaptions(imgObj);
    newImageObj = prepDropDowns(imgObj);
    return newImageObj;
  });
  return renderValue;
}

function toDataObject (valueObject, object) {
  // For now, this is a special case taken care of in the App
  return object;
}

registerDataType(DATA_TYPE_NAME, {toRenderString, toDataObject});

const DragHandle = SortableHandle(() => {
  return (
    <div className='reactBasicTemplateEditor-ImageContainer-imageListItemDragHandle'>
      <div/><div/><div/>
    </div>);
});

const SortableUserImage = SortableElement( class UserImage extends React.Component {
  constructor (props) {
    super(props);
    this.handleCaptionChange = this.handleCaptionChange.bind(this);
    this.handleDropDownChange = this.handleDropDownChange.bind(this);
    this.handleChangeImage = this.handleChangeImage.bind(this);
    this.handleRemoveImage = this.handleRemoveImage.bind(this);
  }
  handleCaptionChange (val, index) {
    this.props.onCaptionChange({
      imageId: this.props.item.id,
      captionIndex: index,
      newValue: val
    });
  }

  handleDropDownChange (index, name, value) {
    this.props.onDropDownChange({
      imageId: this.props.item.id,
      dropDownIndex: index,
      newValue: value
    })
  }

  handleChangeImage () {
    this.props.onChangeImage(this.props.item.id);
  }

  handleRemoveImage () {
    this.props.onRemoveImage(this.props.item.id);
  }

  renderCaptions () {
    if (this.props.item.captions === undefined) {
      return null;
    }
    const captions = this.props.item.captions.map((capObj, i) => {
      return (
        <CaptionInput
          key={ i }
          type='text'
          settings={ capObj.settings }
          data-index={ i }
          value={ capObj.value }
          label={ capObj.label }
          onChange={ this.handleCaptionChange }
          />
      );
    });
    return captions;
  }

  renderButtons () {
    let swapButton, removeButton;
    if (toType(this.props.onChangeImage) !== 'function') {
      swapButton = null;
    } else {
      swapButton = (<button
        key={'swap'} className="reactBasicTemplateEditor-ImageContainer-swapImageButton"
        type="button"
        onClick={ this.handleChangeImage }>
        Swap Image
      </button>);
    }

    if (this.props.onRemoveImage) {
      removeButton = (<button
        key={'remove'} className="reactBasicTemplateEditor-ImageContainer-removeImageButton"
        type="button"
        onClick={ this.handleRemoveImage }>
        Remove
      </button>);
    } else {
      removeButton = null;
    }

    if (swapButton || removeButton) {
      return (
        <div className="reactBasicTemplateEditor-ImageContainer-actionButtons">
          {[swapButton, removeButton]}
        </div>
      );
    } else {
      return null;
    }
  }

  renderDropDowns() {
    if (forceArray(this.props.item.dropDowns).length < 1) return null;
    return this.props.item.dropDowns.map((dropDown, i) => {
      return <ImageDropDown key={ i }
        index={ i }
        fieldName={ dropDown.label }
        value={ dropDown.value }
        onChange={ this.handleDropDownChange }
        options={ dropDown.options }/>
    })
  }

  render () {
    const captions = this.renderCaptions();
    const buttons = this.renderButtons();
    const dropDowns = this.renderDropDowns();
    return (
      <div className='reactBasicTemplateEditor-ImageContainer-imageListItem'>
        <div className='reactBasicTemplateEditor-ImageContainer-imageListItemData'>
          <img src={ THUMBNAIL_URL_PREFIX + this.props.item.file } />
          <div className='reactBasicTemplateEditor-ImageContainer-imageListItemDataChangers'>
            { buttons }
            { captions }
            { dropDowns }
          </div>
        </div>
        <DragHandle />
      </div>
    );
  }
});

const SortableListOfImages = SortableContainer( function ListOfImages (props) {
  return (
    <div>
      {props.items.map((value, index) =>
        <SortableUserImage
          key={`item-${index}`}
          dropDownsSettings={ props.dropDownsSettings }
          onDropDownChange={ props.onDropDownChange }
          onCaptionChange={ props.onCaptionChange }
          index={index}
          onChangeImage={ props.onChangeImage }
          onRemoveImage={ props.onRemoveImage }
          item={value} />
      )}
    </div>
  );
}, {transitionDuration: 0} );

const ImageSelection = (props) => {
  let imageList = props.imageBank.map((image, i) => {
    return <img src={ THUMBNAIL_URL_PREFIX + image } key={ i } onClick={ () => { props.onChooseImage(image) }}/>;
  });
  return (
    <div className="reactBasicTemplateEditor-ImageContainer-imageBank">
      <p className="reactBasicTemplateEditor-ImageContainer-imageBankInstructions">Select a new image</p>
      { imageList }
    </div>
  );
};

class ImageContainer extends React.Component {
  constructor (props) {
    super(props);
    let imagesState = this.getImagesStateFromImages(this.props.images);
    this.state = {modalIsOpen: false, images: imagesState};

    this.handleSortEnd = this.handleSortEnd.bind(this);
    this.handleSortStart = this.handleSortStart.bind(this);
    this.handleCaptionChange = this.handleCaptionChange.bind(this);
    this.handleDropDownChange = this.handleDropDownChange.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
    this.handleChangeImage = this.handleChangeImage.bind(this);
    this.handleRemoveImage = this.handleRemoveImage.bind(this);
    this.handleAddImage = this.handleAddImage.bind(this);
    this.handleReplaceImage = this.handleReplaceImage.bind(this);

  }

  getImagesStateFromImages (images) {
    images = images || [];
    return images.map((val, i) => {
      val.id = i;
      return val;
    });
  }

  handleSortEnd ({oldIndex, newIndex}) {
    let newArray = arrayMove(this.props.images, oldIndex, newIndex);
    this.updateImages(newArray);
    enableTextSelectionOnTheWholeBody();
  }

  handleSortStart () {
    disableTextSelectionOnTheWholeBody();
  }

  componentWillReceiveProps (newProps) {
    if (newProps.images === this.props.images) return;
    this.setState({images: this.getImagesStateFromImages(newProps.images)});
  }

  handleCaptionChange (newCaptionObject) {
    let newArray = this.props.images.map(image => {
      if (image.captions === undefined) {
        // be sure there is an array
        image.captions = [];
      }
      if (image.id === newCaptionObject.imageId) {
        image.captions[newCaptionObject.captionIndex].value = newCaptionObject.newValue;
      }
      return image;
    });
    this.updateImages(newArray);
  }

  handleDropDownChange (newDropDownObject) {
    let newArray = this.props.images.map(image => {
      if (image.dropDowns === undefined) {
        // be sure there is an array
        image.dropDowns = [];
      }
      if (image.id === newDropDownObject.imageId) {
        image.dropDowns[newDropDownObject.dropDownIndex].value = newDropDownObject.newValue;
      }
      return image;
    });
    this.updateImages(newArray);
  }

  updateImages (newArrayOfImages) {
    ContainerActions.changeContainer(
      DATA_TYPE_NAME,
      this.props.fieldName,
      {containedImages: newArrayOfImages}
    );
  }

  openModal () {
    this.setState({modalIsOpen: true})
  }

  handleCloseModal () {
    this.setState({modalIsOpen: false})
  }

  handleChangeImage (oldImageId) {
    this.setState({imageIdToReplace: oldImageId}, this.openModal)
  }

  handleRemoveImage (removeImageId) {
    let newImageArray = clone(this.state.images).filter(image => {
      if (image.id === removeImageId) return false;
      return true;
    });

    this.updateImages(newImageArray);
  }

  wipeCaptions (imageObj) {
    if (imageObj.captions) {
      imageObj.captions = imageObj.captions.map(caption => {
        caption.value = '';
        return caption;
      });
    }
    return imageObj;
  }

  handleAddImage () {
    let newImageArray = this.state.images.concat(
      this.wipeCaptions(clone(this.state.images[0]))
    )
    this.updateImages(newImageArray);
  }

  handleReplaceImage (incomingImage) {
    let outgoingId = this.state.imageIdToReplace;

    let newImageArray = clone(this.state.images).map(image => {
      if (image.id === outgoingId) {
        image.file = incomingImage;
      }
      return image;
    });

    this.updateImages(newImageArray);
    this.handleCloseModal();
  }

  deriveCaptionsSettings (captionsSettingsArr) {
    if (toType(captionsSettingsArr) !== 'array') return captionsSettingsArr;
    return captionsSettingsArr.map(val => {
      if (toType(val) === 'string') return {label: val};
      if (toType(val) === 'object') return val;
      throw new Error(`The value passed in to the captionsDirective should be an array containing strings or objects, '${toType(val)}' given.`);
    })
  }

  shouldAllowRemove () {
    let min = TemplateSpecificationsStore.getSpec('minImages');
    return this.state.images.length > min;
  }

  shouldAllowAdd () {
    let max = TemplateSpecificationsStore.getSpec('maxImages');
    return this.state.images.length < max;
  }

  renderImageList () {
    const images = this.state.images;
    const changeImageFunc = (this.props.imageBank.length > 1) ? this.handleChangeImage : null;
    const removeImageFunc = (this.shouldAllowRemove()) ? this.handleRemoveImage : null;
    const captionsSettings = this.deriveCaptionsSettings(this.props.captions);
    const dropDownsSettings = this.props.dropDowns;
    return (
      <SortableListOfImages
        items={ images }
        onSortEnd={ this.handleSortEnd }
        captionsSettings={ captionsSettings }
        dropDownsSettings={ dropDownsSettings }
        onCaptionChange={ this.handleCaptionChange }
        onDropDownChange={ this.handleDropDownChange }
        useDragHandle={ true }
        onChangeImage={ changeImageFunc }
        onRemoveImage={ removeImageFunc }
        onSortStart={ this.handleSortStart }
      />
    );
  }

  render () {
    if (this.state.images.length === 0) return null;
    const imageList = this.renderImageList();
    const explanation = this.state.images.length > 1 ?
      'Change the order of the images by dragging up and down with the handles on the right.':'';
    return (
      <div className="reactBasicTemplateEditor-ImageContainer">
        <div className="reactBasicTemplateEditor-ImageContainer-explainer">
          { explanation }
        </div>
        { imageList }
        {this.shouldAllowAdd() ? (<button
          className="reactBasicTemplateEditor-ImageContainer-addImageButton"
          type="button"
          onClick={this.handleAddImage}>
          Add Image
        </button>): null}

        <Modal
          isOpen={this.state.modalIsOpen}
          onRequestClose={this.handleCloseModal}
          contentLabel="Choose a replacement image">

          <ImageSelection onChooseImage={ this.handleReplaceImage } imageBank={ this.props.imageBank } />
          <button className="reactBasicTemplateEditor-ImageContainer-modalCancel"
            onClick={ this.handleCloseModal } type="button">
            Cancel
          </button>
        </Modal>
      </div>
    );
  }

}

ImageContainer.propTypes = {
  images: PT.array,
  imageBank: PT.array,
  fieldName: PT.string.isRequired
}

ImageContainer.defaultProps = {
  images: [],
  imageBank: []
}

export default ImageContainer;
