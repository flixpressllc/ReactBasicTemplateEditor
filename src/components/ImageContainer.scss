@import "../styles/base-includes";

.reactBasicTemplateEditor-ImageContainer {
  padding: 0 1em;
}

.reactBasicTemplateEditor-ImageContainer-imageListItem {
  @extend %base-container-styles;
  @include clearfix;
  background-color: rgba(150,150,150,0.5);
  padding: 10px;
  margin: 10px 0;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.reactBasicTemplateEditor-ImageContainer-imageListItemDragHandle {
  cursor: move;
  display: flex;
  flex-direction: column;
  justify-content: center;

  > div {
    width: 1.5em;
    height: 5px;
    background-color: rgba(0,0,0,0.2);
    border: 1px solid rgba(255,255,255, 0.6);
    border-left: none;
    border-top: none;
    margin: 2px;
  }
}

.reactBasicTemplateEditor-ImageContainer-modalCancel {
  @extend %button;
  float: right;
}

.reactBasicTemplateEditor-ImageContainer-imageListItemChangeButton {
  font-size: 0.75em;
  @extend %button;
  padding: 0.35em 0.5em;
}

.reactBasicTemplateEditor-ImageContainer-imageListItemDataChangers {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.reactBasicTemplateEditor-ImageContainer-explainer {
  @extend %explanation-text;
}
