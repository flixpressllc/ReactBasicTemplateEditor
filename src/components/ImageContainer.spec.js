import React from 'react';
import { mount } from 'enzyme';
import ImageContainer from './ImageContainer';
import { isObject } from 'happy-helpers';

jest.mock('../actions/ContainerActions');
const FakeContainerActions = require('../actions/ContainerActions');

const swapButtonSelector = '.reactBasicTemplateEditor-ImageContainer-swapImageButton';
const removeButtonSelector = '.reactBasicTemplateEditor-ImageContainer-removeImageButton';

// <ImageContainer
//   images={ [ arrayOfImages ] }
//   onUpdateImages={ () => {} }
// />

function getSettings ( overrides ) {
  overrides = isObject(overrides) ? overrides : {};
  let defaults = {
    images: [],
    imageBank: [],
    fieldName: 'myImageContainer'
  };
  return Object.assign(defaults, overrides);
}

describe('ImageContainer', () => {
  it('renders without crashing', () => {
    expect(() => {
      mount(<ImageContainer {...getSettings()}/>);
    }).not.toThrow();
  });

  it('displays images', () => {
    let settings = {
      images: [
        {file: 'toast.jpg'},
        {file: 'coffee.jpg'}
      ]
    };
    let component = mount(<ImageContainer {...getSettings(settings)}/>);
    expect(component.find('img').length).toEqual(2);
  });

  it('displays only non-bank images to start', () => {
    let settings = {
      images: [
        {file: 'toast.jpg'},
        {file: 'coffee.jpg'}
      ],
      imageBank: [
        'toast.jpg',
        'coffee.jpg',
        'eggs.jpg',
        'milk.jpg'
      ]
    };
    let component = mount(<ImageContainer {...getSettings(settings)}/>);
    expect(component.find('img').length).toEqual(2);
  });

  describe('remove button', () => {
    it('calls updateImages omitting the removed image', () => {
      let settings = {
        images: [
          {file: 'toast.jpg'},
          {file: 'coffee.jpg'}
        ]
      };
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      component.find(removeButtonSelector).at(0).simulate('click');

      expect(FakeContainerActions.changeContainer).toHaveBeenLastCalledWith('userImageChooser', 'myImageContainer', {'containedImages': [{'file': 'coffee.jpg', 'id': 1}]});
    });
  });

  describe('swap button', () => {
    it('displays the image bank when the swap button is clicked', () => {
      let settings = {
        images: [
          {file: 'toast.jpg'},
          {file: 'coffee.jpg'}
        ],
        imageBank: [
          'toast.jpg',
          'coffee.jpg',
          'eggs.jpg',
          'milk.jpg'
        ]
      };
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      component.find(swapButtonSelector).at(0).simulate('click');
      let renderedImages = component.render().find('img');

      expect(renderedImages.length).toEqual(4);
      settings.imageBank.map((val, i) => {
        expect(renderedImages.eq(i).attr('src')).toContain(val);
      });
    });

    it('swaps out the image when a new one is chosen', () => {
      let settings = {
        images: [
          {file: 'toast.jpg'}
        ],
        fieldName: 'myField',
        imageBank: [
          'toast.jpg',
          'coffee.jpg',
          'eggs.jpg',
          'milk.jpg'
        ]
      };
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      component.find(swapButtonSelector).at(0).simulate('click');
      component.find('img').at(2).simulate('click');

      expect(FakeContainerActions.changeContainer).toHaveBeenLastCalledWith('userImageChooser', 'myField', {'containedImages': [{'file': 'eggs.jpg', 'id': 0}]});
    });

    describe('when the image bank has 1 image', () => {
      it('will not display the Change Image button', () => {
        let settings = {
          images: [
            {file: 'toast.jpg'}
          ],
          imageBank: [
            'toast.jpg'
          ]
        };
        const component = mount(<ImageContainer {...getSettings(settings)}/>);

        expect(component.find(swapButtonSelector).length).toEqual(0);
      });
    });

  });


  describe('captions:', () => {
    it('will show the correct number of caption fields', () => {
      let settings = {
        images: [ {file: 'toast.jpg', captions:['','','']}, {file: 'coffee.jpg', captions:['','','']} ],
        captions: [ 'one','two','three' ]
      }
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      expect(component.find('input').length).toEqual(6);
    });

    it('will label the caption fields properly', () => {
      let settings = {
        images: [ {file: 'toast.jpg', captions:['','','']}, {file: 'coffee.jpg', captions:['','','']} ],
        captions: [ 'one','two','three' ]
      }
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      expect(component.find('input').get(0).placeholder).toEqual('Optional one');
      expect(component.find('input').get(1).placeholder).toEqual('Optional two');
      expect(component.find('input').get(2).placeholder).toEqual('Optional three');
    });

    it('will allow for an object with settings as the main captions directive', () => {
      let settings = {
        images: [ {file: 'toast.jpg', captions:['','','']}, {file: 'coffee.jpg', captions:['','','']} ],
        captions: [
          'one',
          { label: 'two', settings: {maxCharacters: 3} },
          'three'
        ]
      }
      const component = mount(<ImageContainer {...getSettings(settings)}/>);

      expect(component.find('input').get(1).placeholder).toEqual('Optional two');
    });

    describe('filter options', () => {
      describe('maxCharacters', () => {
        it('allows only last character typed if set to 1', () => {
          let settings = {
            images: [ {file: 'toast.jpg', captions:['','','']}, {file: 'coffee.jpg', captions:['','','']} ],
            captions: [
              'one',
              { label: 'two', settings: {maxCharacters: 1} },
              'three'
            ]
          }
          const fakeEvent = {target:{value:'abcdefg'}};
          const component = mount(<ImageContainer {...getSettings(settings)}/>);

          component.find('input').at(1).simulate('change', fakeEvent);

          expect(FakeContainerActions.changeContainer).toMatchSnapshot()
        });

        it('allows only first n characters when set to n > 1', () => {
          let settings = {
            images: [ {file: 'toast.jpg', captions:['','','']}, {file: 'coffee.jpg', captions:['','','']} ],
            captions: [
              'one',
              { label: 'two', settings: {maxCharacters: 3} },
              'three'
            ]
          }
          const fakeEvent = {target:{value:'abcdefg'}};
          const component = mount(<ImageContainer {...getSettings(settings)}/>);

          component.find('input').at(1).simulate('change', fakeEvent);

          expect(FakeContainerActions.changeContainer).toMatchSnapshot()
        });
      });
    });
  });

});
