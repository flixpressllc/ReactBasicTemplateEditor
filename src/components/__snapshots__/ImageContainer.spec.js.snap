// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageContainer captions: filter options maxCharacters allows only first n characters when set to n > 1 1`] = `[Function]`;

exports[`ImageContainer captions: filter options maxCharacters allows only last character typed if set to 1 1`] = `[Function]`;

exports[`ImageContainer the toRenderString function behaves consistently when there are captions and drop downs 1`] = `
Array [
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "something familair",
      },
      Object {
        "label": "Middle Text",
        "value": "somet",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [
      Object {
        "label": "Which Kid?",
        "value": "1",
      },
    ],
    "file": "DonDentonAdmin_money.jpg",
    "id": 0,
  },
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "",
      },
      Object {
        "label": "Middle Text",
        "value": "",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [
      Object {
        "label": "Which Kid?",
        "value": "2",
      },
    ],
    "file": "DonDentonAdmin_hammer.jpg",
    "id": 1,
  },
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "",
      },
      Object {
        "label": "Middle Text",
        "value": "",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [
      Object {
        "label": "Which Kid?",
        "value": "1",
      },
    ],
    "file": "DonDentonAdmin_tree.jpg",
    "id": 2,
  },
]
`;

exports[`ImageContainer the toRenderString function behaves consistently when there are captions without dropDowns 1`] = `
Array [
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "something familair",
      },
      Object {
        "label": "Middle Text",
        "value": "somet",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_money.jpg",
    "id": 0,
  },
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "",
      },
      Object {
        "label": "Middle Text",
        "value": "",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_hammer.jpg",
    "id": 1,
  },
  Object {
    "captions": Array [
      Object {
        "label": "Top Text",
        "value": "",
      },
      Object {
        "label": "Middle Text",
        "value": "",
      },
      Object {
        "label": "Bottom Text",
        "value": "",
      },
    ],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_tree.jpg",
    "id": 2,
  },
]
`;

exports[`ImageContainer the toRenderString function behaves consistently when there are no captions or dropDowns 1`] = `
Array [
  Object {
    "captions": Array [],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_money.jpg",
    "id": 0,
  },
  Object {
    "captions": Array [],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_hammer.jpg",
    "id": 1,
  },
  Object {
    "captions": Array [],
    "dropDowns": Array [],
    "file": "DonDentonAdmin_tree.jpg",
    "id": 2,
  },
]
`;
